internal/api/auth/ldap.go

```diff
@@ -1,41 +1,212 @@
+		baseDN:    baseDN,
+		reqCh:     make(chan authRequest),
 	}
 }
 
+func SetLDAPDefaultTimeout(timeout time.Duration) {
+	ldap.DefaultTimeout = timeout //nolint:reassign  // we want to customize the default timeout
+}
+
+// SetMaxConnectionLifetime sets the maximum lifetime of a connection.
+//
+// The maximum lifetime is the maximum amount of time a connection may be reused for.
+// This is not a guarantee, as the connection may have been closed by the server before reaching that timer.
+func (l *LDAPAuth) SetMaxConnectionLifetime(maxConnectionLifetime time.Duration) {
+	l.maxConnectionLifetime = maxConnectionLifetime
+}
+
+// StartAuthenticationWorkers starts a pool of workers that will handle the authentication requests.
+func (l *LDAPAuth) StartAuthenticationWorkers(ctx context.Context, workersCount int) error {
+	if workersCount <= 0 {
+		return fmt.Errorf("'WorkersCount' must be greater than 0: %d", workersCount)
+	}
+	for i := 0; i < workersCount; i++ {
+		go l.spawnConnectionWorker(ctx)
+	}
+	return nil
+}
 func (l *LDAPAuth) AuthenticateUser(username string, password string) bool {
-	if err := l.ldapClient.Bind(l.bindDN, l.password); err != nil {
+	req := authRequest{
+		username: username,
+		password: password,
+		authResp: make(chan bool),
+	}
+	l.reqCh <- req
+	return <-req.authResp
+}
+
+func (l *LDAPAuth) spawnConnectionWorker(ctx context.Context) {
+	const maxRetry = 1
+	var conn *ldap.Conn
+	var err error
+	tick := time.NewTicker(l.maxConnectionLifetime)
+
+	for {
+		select {
+		case req := <-l.reqCh:
+			auth := false
+			retry := 0
+			for retry <= maxRetry+1 {
+				retry++
+				log.Debug().Msgf("worker LDAP authentication attempt number %d", retry)
+				// (re)connect if needed
+				if conn == nil || conn.IsClosing() {
+					log.Debug().Msg("LDAP connection is closed, reconnecting")
+					conn, err = l.connect()
+					if err != nil {
+						log.Error().Err(err).Msg("worker LDAP reconnection failed")
+						req.authResp <- false
+						break
+					}
+				}
+
+				// bind with the user credentials
+				var connState connectionStatus
+				ctxTimeout, cancel := context.WithTimeoutCause(ctx, ldap.DefaultTimeout, ErrLDAPTimeout)
+				auth, connState = l.authenticateWithTimeout(ctxTimeout, conn, req.username, req.password)
+				cancel()
+
+				if connState == connectionClosed {
+					log.Debug().Msg("LDAP connection was closed by the server, closing on client side")
+					if err := conn.Close(); err != nil {
+						log.Error().Err(err).Msg("connection was closed by the server but failed to close on client side")
+					}
+				} else {
+					// LDAP connection is still up, we accept the authentication result
+					log.Debug().Msg("auth response valid")
+					break
+				}
+			}
+
+			log.Debug().Msgf("worker LDAP authentication attempt number %d, result: %t", retry, auth)
+
+			req.authResp <- auth
+			tick.Reset(l.maxConnectionLifetime)
+
+		case <-tick.C:
+			// close connection if no request has been made for a minute
```

Comment:
nitpick: don't precise the time here as it depends on the maxConnectionLifetime set

Reply:
done